### PR TITLE
`error-stack`: introduction of color modes

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

This implements the notion of color modes, determining if colors and which type of coloring/styling are requested.

~~This enables users to override the preferred mode while preserving the convenience of environment variables like `NO_COLOR` and `FORCE_COLOR`.~~ This has been moved into an example.

The available color modes are `Color`, `None`, and `Emphasis`. This mode is available for hooks so that they can easily adapt. A perfect example of this is the location hook, which now will use the `mode` to either print in bright black or italic.

~~If no color mode is set, the mode is automatically determined using `owo-colors` using `supports-colors` and the specified override.~~ If no color mode is set, the mode is set to `Color::Emphasis` if the `color` feature (opt-in) is enabled.

> **Note:** as a drive-by, `supports-colors` is now only enabled on `std`, `supports-color` implies `std` (not as a feature, found this by accident), which meant that no-std colors would fail, even tho they shouldn't.

## 🔗 Related links

* #1782

## 🔍 What does this change?

* `Report::format_color_mode_preference`
* `HookContext::mode()` is now available

## 📜 Does this require a change to the docs?

Yes, the new `ColorMode` needs to be explained. This is already done.

We might want to add additional explanation/info in `lib.rs`

## 🏃 Next Steps

1. Ideally we'd temporarily set the mode from owo-colors, if possible, this needs https://github.com/jam1garner/owo-colors/pull/84

## 🛡 What tests cover this?

* Existing tests
* New snapshot tests

## 📹 Demo

image

image

### CODE REVIEW

**Comment by indietyp** [comment]
discussed online with @TimDiekmann, rework in progress which will remove `pretty-print`, introduce a `Charset` and `ColorMode`

File: packages/libs/error-stack/tests/snapshots/doc/fmt__charset_ascii.snap
Comment (TimDiekmann) [inline packages/libs/error-stack/tests/snapshots/doc/fmt__charset_ascii.snap:2]: I'm not sure but do we want to add the `"at "` here as well? If so, we could also remove the whitespace between `-` and `at` again I guess.

File: packages/libs/error-stack/tests/snapshots/doc/fmt__preference_color.snap
Comment (TimDiekmann) [inline packages/libs/error-stack/tests/snapshots/doc/fmt__preference_color.snap:2]: I have not noticed that before as the output is the same, but do we emit unneeded spans here, or is this due to the conversion to HTML-codes?

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/charset.rs:35]: `AtomicOverride` is very similar to `ColorPreference`, Is it possible to reuse the implementation?

**Review by TimDiekmann — COMMENTED** [review]
Looks good to me, thanks @indietyp! 🎉 

I have a few comments which should be addressed before we merge it, but nothing, which is really blocking the merge.

File: packages/libs/error-stack/tests/snapshots/doc/fmt__preference_color.snap
Comment (indietyp) [reply packages/libs/error-stack/tests/snapshots/doc/fmt__preference_color.snap]: Great catch. After looking at the code, we actually emit the ANSI codes multiple times, while in reality, there's no real change. It is considered poor etiquette. Fixed in [8f5918b](https://github.com/hashintel/hash/pull/1800/commits/8f5918bfa5cec7b666cf544ed74ac8663a6ce85d)

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/charset.rs]: fixed in [de12b05](https://github.com/hashintel/hash/pull/1800/commits/de12b05ff5dae1663f87c3009307f929b38355bf)

File: packages/libs/error-stack/tests/snapshots/doc/fmt__charset_ascii.snap
Comment (indietyp) [reply packages/libs/error-stack/tests/snapshots/doc/fmt__charset_ascii.snap:2]: 🤔 I am unsure. We could if that is the case, would it be "at <color>LOCATION</color>" where color is either color or emphasis or another way?

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:53]: The behavior using `Option` feels weird to me. We either have `None` or `Some(None)`. I'd prefer to always return a value, if none was set before we should just return our default value.

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:42]: Does this mean we cannot store `ColorMode` when the `"color"` feature is not enabled?

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:51]: Why this requires a feature?

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:53]: Why do we have two implementations of `load` with different semantics?

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:149]: Why this is gated behind a feature flag? I expect that downstream binaries want to set `ColorMode::None` or `ColorMode::Emphasis` and downstream libraries can pick up those values, this has nothing to do on our dependency on `owo-colors`.

File: packages/libs/error-stack/src/lib.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/lib.rs:473]:
```suggestion
```

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
I found an issue regarding the feature flags, the behavior for color support should be similar to `charset` and the `color` feature enables the usage of colored output e.g. for the location.

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs:42]: Yes, I tied the `color` feature to the ability to enable color, but after thinking about it, that's a mistake. We should not gate that behavior.

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [f83228f](https://github.com/hashintel/hash/pull/1800/commits/f83228f7d0c91828e09c642bd916119794d17402)

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [f83228f](https://github.com/hashintel/hash/pull/1800/commits/f83228f7d0c91828e09c642bd916119794d17402)

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [f83228f](https://github.com/hashintel/hash/pull/1800/commits/f83228f7d0c91828e09c642bd916119794d17402)

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [f83228f](https://github.com/hashintel/hash/pull/1800/commits/f83228f7d0c91828e09c642bd916119794d17402)

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [f83228f](https://github.com/hashintel/hash/pull/1800/commits/f83228f7d0c91828e09c642bd916119794d17402)

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/charset.rs]: Can we `derive` it instead?

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs]: This feels wrong, especially when in mind, that we are going to remove the `color` feature in a follow-up. Can we just derive `Default` and set it to `Emphasis`?

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/charset.rs:10]: We should derive `PartialEq`, `Eq`, and `Hash` as well

File: packages/libs/error-stack/src/fmt/color.rs
Comment (TimDiekmann) [inline packages/libs/error-stack/src/fmt/color.rs:10]: We should derive `Hash` as well

Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [a727237](https://github.com/hashintel/hash/pull/1800/commits/a727237a55d6a1fdcad4abc805e546800833dd14)

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/charset.rs]: fixed in [a727237](https://github.com/hashintel/hash/pull/1800/commits/a727237a55d6a1fdcad4abc805e546800833dd14)

File: packages/libs/error-stack/src/fmt/color.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/color.rs]: fixed in [a727237](https://github.com/hashintel/hash/pull/1800/commits/a727237a55d6a1fdcad4abc805e546800833dd14)

File: packages/libs/error-stack/src/fmt/charset.rs
Comment (indietyp) [reply packages/libs/error-stack/src/fmt/charset.rs]: fixed in [a727237](https://github.com/hashintel/hash/pull/1800/commits/a727237a55d6a1fdcad4abc805e546800833dd14)

File: packages/libs/error-stack/CHANGELOG.md
Comment (TimDiekmann) [inline packages/libs/error-stack/CHANGELOG.md:23]: Just as a reminder, that we have to tweak this entry when removing the feature again 🙂

**Review by TimDiekmann — DISMISSED** [review]
Thanks @indietyp, a nice, concise interface! 🙂 

Too bad, that `rustdoc` does not render `#[default]`.

**Review by TimDiekmann — DISMISSED** [review]

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**
